### ISSUE TITLE
Building Model Layers

### ISSUE THREAD

**Issue by hmustafajr** [hmustafajr/PyTorch-Fundamentals#4]
What is a neural network?
A neural network is a collection of neurons that are connected by layers. Each neuron is a small computing unit that performs simple calculations to collectively solve a problem. Neurons are organized in 3 types of layers:

Inputs
Hiddens
Outputs
The hidden and output layers contain a number of neurons. Neural networks mimic the way our brains process information.

Components of a neural network

An activation function determines whether a neuron should be activated or not. The computation that happens in a neural network include applying an activation function. If a neuron activates, then it means the input is important. There are different kinds of activation functions. The choice of which activation function to use depends on what you want the output to be. Another important role of an activation function is to add non-linearity to the model.

Binary is used to set an output node to 1 if the function result is positive and 0 if the function result is zero or negative. f(x)={ 0, 1, if x<0 if x≥0

Sigmoid is used to predict the probability of an output node being between 0 and 1, for classification use cases. f(x) = (1) / (1+e^-x)

Tanh is used to predict whether an output node is between 1 and -1, for classification use cases. f(x) = (e^x - e^-x) / (e^x + e^-x)

ReLU (rectified linear activation function) is used to set the output node to 0 if the function result is negative and keeps the result value if the result is a positive value. f(x)={ 0, x, if x<0 if x≥0

Weights influence how close the output of our network is to the expected output value. As an input enters the neuron, it gets multiplied by a weight value and the resulting output is either observed, or passed to the next layer in the neural network. Weights for all neurons in a layer are organized into one tensor.

Bias makes up the difference between the activation function's output and its intended output. A low bias suggest that the network is making more assumptions about the form of the output, whereas a high bias value makes fewer assumptions about the form of the output.

We can say that an output Y of a neural network layer with weights W and bias B is computed as summation of the inputs multiplied by the weights plus the bias.

x=∑(weights∗inputs)+bias, where f(x) is the activation function

Neural networks are comprised of layers and modules that perform operations on data. The **torch.nn** namespace provides all the building blocks needed to build your own neural network. Every module in PyTorch subclasses **nn.Module**. A neural network itself a module that consists of other modules (layers). This nested structure allows for building and managing complex architectures easily.

In the next section, we'll build a neural network to classify images in thee FashionMNIST dataset.

[Neural Network - Build](https://github.com/hmustafajr/Python-Toolbox/blob/main/Neural%20Network%20-%20Build.py)

**Get Hardware device for training**
We want to be able to train our model on a hardware accelerator like a GPU, if one is available. Let's check to see whether **torch.cuda** is available; if not, we'll continue to use the CPU.

[Get Hardware Device](https://github.com/hmustafajr/Python-Toolbox/blob/main/Hardware%20Device%20-%20Get.py)

**Defining Class**
We define our neural network by subclassing _nn.Module_, and initialize the neural network layers in **____init____**. Every **nn.Module** subclass implements the operations on input data in the **forward** method.

Our neural network is composed of the following:
- The input layer with 28x28 or 784 features/pixels.
- The first linear module takes the input 784 features and transforms it to a hidden layer with 512 feature.
- The ReLU activation function will be applied in the transformation.
- The second linear module takes 512 features as input from the first hidden layer and transforms it to the next hidden layer with 512 features.
- The ReLU activation function will be applied in the transformation.
- The third linear module takes 512 features as input from the second hidden layer and transforms those features to the output layer within 10, which is the number of classes.
- The ReLU activation function will be applied in the transformation.

[Definifng Class](https://github.com/hmustafajr/Python-Toolbox/blob/main/Defining%20Class.py)
We create an instance of **NeuralNetwork**, and move it to the **device** and print its structure.

To use the model, we pass it the input data. This executes the model's **forward**, along with some background operations. However, do not call **model.forward()** directly! Calling the model on the input returns a 10-dimentional tensor with raw predicted values for each class.

We get the prediction densities by passing it through an instance of the **nn.Softmax**.

**Weight & Bias**
The **nn.Linear** module randomly initializes the _weights_ and _bias_ for each layer and internally stores the values in Tensors.
`print(f"First Linear weights: {model.linear_relu_stack[0].weight} \n")
print(f"First Linear biases: {model.linear_relu_stack[0].bais} \n")`

**Model Layers**
Let's break down the layers in the FashionMNIST model. To picture it, we will take a sample mini batch of 3 images sized 28x28 and see what happens when we pass it through the network.
`input_image = torch.rand(3,28,28)`
`print(input_image.size())`

[Model Layers](https://github.com/hmustafajr/Python-Toolbox/blob/main/Model%20Layers.py)

**Flatten**
We initialize the nn.Flatten later to convert each 2 dimensional 28x28 image into a contiguous array of 784 pixel values, because the minibatch dimension (at dim=0) is maintained. Each of the pixels are passed to the input layer of the neural network.
`flatten = nn.Flatten()`
`flat_image = flatten(input_image)`
`print(flat_image.size())`

[Flatten](https://github.com/hmustafajr/Python-Toolbox/blob/main/Flatten.py)

**nn,Linear**
The linear layer is a module that applies a linear transformation on the input using its stored weights and biases. The grayscale value of each pixel in the input layer will be connected to neurons in the hidden layer for calculations. The calculation used for the transformation is _weight * input + bias_.

`layer1 = nn.Linear(in_features=28*28, out_features=20`
`hidden1 = layer1(flat_image)`
`print(hidden1.size())`

[Linear](https://github.com/hmustafajr/Python-Toolbox/blob/main/Linear.py)

**Non-Linear** (ReLU)
Non-linear activations are what create the complex mappings between the model's inputs and outputs. They are applied after linear transformations to introduce nonlinearity, helping neural networks learn a wide variety of phenomena. In this model, we use _nn.ReLU__ between our linear layers, but there's other activations to introduce non-linearity in your model.

The ReLU activation function takes the output from the linear layer calculations and replaces the negative values with zeros.

Linear output **x** = **weight * input +bias**.
ReLU: f(x)={ 0, if x<0
                   x, if x≥0

`print(f"Before ReLU: {hidden1}\n\n")`
`hidden1 = nn.ReLU()(hidden1)`
`print(f"After ReLU: {hidden1}")`

[Non-Linear (ReLU)](https://github.com/hmustafajr/Python-Toolbox/blob/main/Non-Linear%20(ReLU).py)

**Sequential**
**nn.Sequential** is an ordered container of modules. The data is passed through all the modules in their defined order. You can use sequential containers to put together a quick network like seq_modules.

`seq_modules = nn.Sequential(`
​`flatten,`
`later1,`
`nn.ReLU(),`
`nn.Linear(20, 10)`
`)`
`input_image = torch.rand(3,28,28)`
`logits = seq_modules(input_image)`

[nn.Sequential](https://github.com/hmustafajr/Python-Toolbox/blob/main/Sequential.py)

**nn.Softmax**
The final linear layer of the neural network returns _logits_ (the raw values in [-_infty_, _infty_]), which are passed to the _nn.softmax_ module. The Softmax activation function is used to calculate the probability of the output from the neural network. It is only used on the output layer of a neural network. The results are scaled to values [0,1] representing the model's predicted densities for each class. The _dim_ parameter indicates the dimension along which the result values must sum to 1. The node with the highest probability predicts the desired output.

`softmax = nn.Softmac(dim=1)`
`pred_probab = softmax(logits)`

[Softmax](https://github.com/hmustafajr/Python-Toolbox/blob/main/Softmax.py)

**Model Parameters**
Many layers inside a neural network are _parameterized_, that is, the layers have associated weights and biases that are optimized during training. Subclassing _nn.Module_ automatically tracks all fields defined inside your model object, and makes all parameters accessible using your model's _parameter()_ or _named_parameters()_ methods.

In this example, we iterate over each parameter, and print its size and a preview of its values.

`print("Model structure: ", model, "\n\n")`
`for name, param in model.named_parameters():`
`print(f"Layer: {name} | Size: {param.size()} | Values : {param[2]} \n")`

[Model Parameters](https://github.com/hmustafajr/Python-Toolbox/blob/main/Model%20Parameters.py)
